File: code/make_users.py
# Purpose:

# Import modules
import arcpy
from arcpy.sa import *

# Settings
arcpy.env.overwriteOutput = True
OUTPUT = "H:/analysis/ca/gis/calwaterusers.gdb/"
FINAL_SHAPEFILES = "H:/analysis/ca/gis/shapefile_output/"
FINAL_TABLES = "H:/analysis/ca/gis/table_output/"

# Input files
RAW_HUC12 = "H:/data/GIS/ca/dwr/CAWBD_Certified/CA_SDE_Extraction_17Dec2008.mdb/WBD/HU12_polygon"
RAW_COUNTY = "H:\data\GIS\ca\ca_atlas\counties\shapefiles\cnty24k09_1_multipart.shp"
RAW_DAU = "H:/data/GIS/ca/dwr/dau_v2_CA105/dau_v2_105.shp"
RAW_DWR_AGENCIES = "H:/data/GIS/ca/dwr/districts/combined/water_agencies.shp"
RAW_ATLAS_FEDERAL = "H:/data/GIS/ca/ca_atlas/water_district_boundaries/federal/WD-WaterUsers.mdb/FederalWaterUsers"
RAW_ATLAS_STATE = "H:/data/GIS/ca/ca_atlas/water_district_boundaries/state/wdst24.shp"
RAW_ATLAS_PRIVATE = "H:/data/GIS/ca/ca_atlas/water_district_boundaries/private/wdpr24.shp"
RAW_MOJAVE = "H:/data/GIS/ca/mojave/Mojave_Water_Agency_Service_Area_Water_Companies_2012.shp"
RAW_CEHTP ="H:/data/GIS/ca/cehtp/all_mod/service_areas.shp"
RAW_CROPMASK = "H:/data/CA/usda_cdl/tif/CMASK_2015_06.tif"
RAW_ATLAS_FEDERAL_TABLE ="H:/data/GIS/ca/ca_atlas/water_district_boundaries/federal/WD-WaterUsers.mdb/Master"

# Set some initial Arc stuff
coordsys_geog = arcpy.SpatialReference('WGS 1984')
coordsys_proj = arcpy.SpatialReference('NAD 1983 UTM Zone 10N')
arcpy.CheckOutExtension("Spatial")

# Program
try:

# Dissolve within layer, into multipart features
print "Dissolving..."
arcpy.Dissolve_management(in_features=RAW_DWR_AGENCIES,
                          out_feature_class=OUTPUT+"agencies_1",
                          dissolve_field="AGENCYUNIQ", statistics_fields="AGENCYNAME FIRST",
                          multi_part="MULTI_PART", unsplit_lines="DISSOLVE_LINES")
arcpy.Dissolve_management(in_features=RAW_ATLAS_FEDERAL,
                          out_feature_class=OUTPUT+"federal_1",
                          dissolve_field="WDNAME",
                          multi_part="MULTI_PART", unsplit_lines="DISSOLVE_LINES")
arcpy.Dissolve_management(in_features=RAW_ATLAS_STATE,
                          out_feature_class=OUTPUT+"swp_1",
                          dissolve_field="WDNAME",
                          multi_part="MULTI_PART", unsplit_lines="DISSOLVE_LINES")
arcpy.Dissolve_management(in_features=RAW_ATLAS_PRIVATE,
                          out_feature_class=OUTPUT+"private_1",
                          dissolve_field="WDNAME",
                          multi_part="MULTI_PART", unsplit_lines="DISSOLVE_LINES")
arcpy.Dissolve_management(in_features=RAW_MOJAVE
                          out_feature_class=OUTPUT+"mojave_1",
                          dissolve_field="Name",
                          multi_part="MULTI_PART", unsplit_lines="DISSOLVE_LINES")
arcpy.Dissolve_management(in_features=RAW_CEHTP,
                          out_feature_class=OUTPUT+"cehtp_1",
                          dissolve_field="pwsid", statistics_fields="pwsname FIRST",
                          multi_part="MULTI_PART", unsplit_lines="DISSOLVE_LINES")
arcpy.Dissolve_management(in_features=RAW_DAU,
                          out_feature_class=OUTPUT+"dau_1",
                          dissolve_field="DAU_CODE", statistics_fields="DAU_NAME FIRST; PSA_CODE FIRST; PSA_NAME FIRST; HR_CODE FIRST; HR_NAME FIRST; PA_NO FIRST",
                          multi_part="MULTI_PART", unsplit_lines="DISSOLVE_LINES")
arcpy.Dissolve_management(in_features=RAW_HUC12,
                          out_feature_class=OUTPUT+"huc8_1",
                          dissolve_field="HUC_8", statistics_fields="",
                          multi_part="MULTI_PART", unsplit_lines="DISSOLVE_LINES")

# Intersect DAU and county
print "Intersecting..."
arcpy.Intersect_analysis(in_features=[OUTPUT+"dau_1", RAW_COUNTY],
                         out_feature_class=OUTPUT+"dauco_1",
                         join_attributes="NO_FID",
                         cluster_tolerance=100)

# Delete unneeded fields
print "Deleting fields..."
arcpy.DeleteField_management(OUTPUT+"dauco_1", "NAME_UCASE")
arcpy.DeleteField_management(OUTPUT+"dauco_1", "FMNAME_PC")
arcpy.DeleteField_management(OUTPUT+"dauco_1", "FMNAME_UC")
arcpy.DeleteField_management(OUTPUT+"dauco_1", "ABBREV")
arcpy.DeleteField_management(OUTPUT+"dauco_1", "ABCODE")
arcpy.DeleteField_management(OUTPUT+"dauco_1", "Shape_Leng")

# Rename fields
print "Renaming fields..."
arcpy.AlterField_management(in_table=OUTPUT+"agencies_1", field="FIRST_AGENCYNAME", new_field_name="username")
arcpy.AlterField_management(in_table=OUTPUT+"federal_1", field="WDNAME", new_field_name="username")
arcpy.AlterField_management(in_table=OUTPUT+"swp_1", field="WDNAME", new_field_name="username")
arcpy.AlterField_management(in_table=OUTPUT+"private_1", field="WDNAME", new_field_name="username")
arcpy.AlterField_management(in_table=OUTPUT+"mojave_1", field="Name", new_field_name="username")
arcpy.AlterField_management(in_table=OUTPUT+"cehtp_1", field="FIRST_pwsNAME", new_field_name="username")
arcpy.AlterField_management(in_table=OUTPUT+"dauco_1", field="FIRST_DAU_NAME", new_field_name="DAU_NAME")
arcpy.AlterField_management(in_table=OUTPUT+"dauco_1", field="FIRST_PSA_CODE", new_field_name="PSA_CODE")
arcpy.AlterField_management(in_table=OUTPUT+"dauco_1", field="FIRST_PSA_NAME", new_field_name="PSA_NAME")
arcpy.AlterField_management(in_table=OUTPUT+"dauco_1", field="FIRST_HR_CODE", new_field_name="HR_CODE")
arcpy.AlterField_management(in_table=OUTPUT+"dauco_1", field="FIRST_HR_NAME", new_field_name="HR_NAME")
arcpy.AlterField_management(in_table=OUTPUT+"dauco_1", field="FIRST_PA_NO", new_field_name="PA_NO")
arcpy.AlterField_management(in_table=OUTPUT+"dauco_1", field="NAME_PCASE", new_field_name="COUNTY_NAME")
arcpy.AlterField_management(in_table=OUTPUT+"dauco_1", field="NUM", new_field_name="COUNTY_CODE")
arcpy.AlterField_management(in_table=OUTPUT+"dauco_1", field="ANSI", new_field_name="COUNTY_ANSI")

# Add a field identifying source file
print "Adding fields..."
arcpy.AddField_management(in_table=OUTPUT+"agencies_1", field_name="source", field_type="TEXT")
arcpy.AddField_management(in_table=OUTPUT+"federal_1", field_name="source", field_type="TEXT")
arcpy.AddField_management(in_table=OUTPUT+"swp_1", field_name="source", field_type="TEXT")
arcpy.AddField_management(in_table=OUTPUT+"private_1", field_name="source", field_type="TEXT")
arcpy.AddField_management(in_table=OUTPUT+"mojave_1", field_name="source", field_type="TEXT")
arcpy.AddField_management(in_table=OUTPUT+"cehtp_1", field_name="source", field_type="TEXT")
arcpy.CalculateField_management(in_table=OUTPUT+"agencies_1", field="source", expression="'agencies'", expression_type="PYTHON_9.3")
arcpy.CalculateField_management(in_table=OUTPUT+"federal_1", field="source", expression="'federal'", expression_type="PYTHON_9.3")
arcpy.CalculateField_management(in_table=OUTPUT+"swp_1", field="source", expression="'swp'", expression_type="PYTHON_9.3")
arcpy.CalculateField_management(in_table=OUTPUT+"private_1", field="source", expression="'private'", expression_type="PYTHON_9.3")
arcpy.CalculateField_management(in_table=OUTPUT+"mojave_1", field="source", expression="'mojave'", expression_type="PYTHON_9.3")
arcpy.CalculateField_management(in_table=OUTPUT+"cehtp_1", field="source", expression="'cehtp'", expression_type="PYTHON_9.3")

# Add a field identifying DAUCOid
arcpy.AddField_management(in_table=OUTPUT+"dauco_1", field_name="dauco_id", field_type="LONG")
arcpy.CalculateField_management(in_table=OUTPUT+"dauco_1", field="dauco_id", expression="!DAU_CODE!.encode()", expression_type="PYTHON_9.3")
arcpy.CalculateField_management(in_table=OUTPUT+"dauco_1", field="dauco_id", expression="!dauco_id!*100 + !COUNTY_CODE!", expression_type="PYTHON_9.3")

# Merge users together (projecting into WGS84, the first input's coordinate system)
print "Merging..."
arcpy.Merge_management([OUTPUT+"cehtp_1",   OUTPUT+"agencies_1",    OUTPUT+"federal_1",
                        OUTPUT+"swp_1",    OUTPUT+"private_1",     OUTPUT+"mojave_1", ],
                       OUTPUT+"users_2")

# Add user unique id
arcpy.AddField_management(in_table=OUTPUT+"users_2", field_name="user_id", field_type="LONG")
arcpy.CalculateField_management(in_table=OUTPUT+"users_2", field="user_id", expression="!OBJECTID!", expression_type="PYTHON_9.3")

# Add centroid (restricting to within shape)
print "Calculating centroids..."
arcpy.AddGeometryAttributes_management(OUTPUT+"users_2", "CENTROID_INSIDE")
arcpy.DeleteField_management(OUTPUT+"users_2", "INSIDE_Z")
arcpy.DeleteField_management(OUTPUT+"users_2", "INSIDE_M")
arcpy.AlterField_management(in_table=OUTPUT+"users_3", field="INSIDE_X", new_field_name="user_x")
arcpy.AlterField_management(in_table=OUTPUT+"users_3", field="INSIDE_Y", new_field_name="user_y")

# Project to meters and calculate area
print "Projecting..."
arcpy.Project_management(in_dataset=OUTPUT+"users_2",
                         out_dataset=OUTPUT+"users_3",
                         out_coor_system=coordsys_proj)
arcpy.Project_management(in_dataset=OUTPUT+"huc8_1",
                         out_dataset=OUTPUT+"huc8_2",
                         out_coor_system=coordsys_proj)
arcpy.Project_management(in_dataset=OUTPUT+"dauco_1",
                         out_dataset=OUTPUT+"dauco_2",
                         out_coor_system=coordsys_proj)

# Add shape areas
print "Calculating areas..."
arcpy.AddGeometryAttributes_management(OUTPUT+"users_3", "AREA", "", "SQUARE_KILOMETERS")
arcpy.AlterField_management(in_table=OUTPUT+"users_3", field="POLY_AREA", new_field_name="totarea")
arcpy.AddGeometryAttributes_management(OUTPUT+"huc8_2", "AREA", "", "SQUARE_KILOMETERS")
arcpy.AlterField_management(in_table=OUTPUT+"huc8_2", field="POLY_AREA", new_field_name="huc8_area")
arcpy.AddGeometryAttributes_management(OUTPUT+"dauco_2", "AREA", "", "SQUARE_KILOMETERS")
arcpy.AlterField_management(in_table=OUTPUT+"dauco_2", field="POLY_AREA", new_field_name="dauco_area")

# Calculate zonal statistics on cropmask and join back
print "Calculating cropmask..."
arcpy.gp.ZonalStatisticsAsTable_sa(OUTPUT+"huc8_2", "HUC_8", RAW_CROPMASK, OUTPUT+"huc8_cropmask", "DATA", "MEAN")
arcpy.gp.ZonalStatisticsAsTable_sa(OUTPUT+"dauco_2", "dauco_id", RAW_CROPMASK, OUTPUT+"dauco_cropmask", "DATA", "MEAN")
arcpy.JoinField_management(in_data=OUTPUT+"huc8_2", in_field="HUC_8",
                           join_table=OUTPUT+"huc8_cropmask", join_field="HUC_8", fields="MEAN")
arcpy.JoinField_management(in_data=OUTPUT+"dauco_2", in_field="dauco_id",
                           join_table=OUTPUT+"dauco_cropmask", join_field="dauco_id", fields="MEAN")
arcpy.AlterField_management(in_table=OUTPUT+"huc8_2", field="MEAN", new_field_name="huc8_pctcrop")
arcpy.AlterField_management(in_table=OUTPUT+"dauco_2", field="MEAN", new_field_name="dauco_pctcrop")

# Project to DD
print "Projecting..."
arcpy.Project_management(in_dataset=OUTPUT+"users_3",
                         out_dataset=OUTPUT+"users_4",
                         out_coor_system=coordsys_geog)
arcpy.Project_management(in_dataset=OUTPUT+"huc8_2",
                         out_dataset=OUTPUT+"huc8_3",
                         out_coor_system=coordsys_geog)
arcpy.Project_management(in_dataset=OUTPUT+"dauco_2",
                         out_dataset=OUTPUT+"dauco_3",
                         out_coor_system=coordsys_geog)

# Export shapefiles
print "Exporting..."
arcpy.FeatureClassToFeatureClass_conversion(in_features=OUTPUT+"huc8_3", out_path=FINAL_SHAPEFILES, out_name="huc8_final")
arcpy.FeatureClassToFeatureClass_conversion(in_features=OUTPUT+"dauco_3", out_path=FINAL_SHAPEFILES, out_name="dauco_final")
arcpy.FeatureClassToFeatureClass_conversion(in_features=OUTPUT+"users_4", out_path=FINAL_SHAPEFILES, out_name="users_final")




# Intersect users with DAUCo and HUC8
print "Intersecting (1/2)..."
arcpy.Intersect_analysis(in_features=[OUTPUT+"users_3", OUTPUT+"huc8_2"],
                         out_feature_class=OUTPUT+"users_huc8_4",
                         join_attributes="NO_FID",
                         cluster_tolerance=1)
print "Intersecting (2/2)..."
arcpy.Intersect_analysis(in_features=[OUTPUT+"users_3", OUTPUT+"dauco_2"],
                         out_feature_class=OUTPUT+"users_dauco_4",
                         join_attributes="NO_FID",
                         cluster_tolerance=1)
print "Dissolving (1/2)..."
arcpy.Dissolve_management(in_features=OUTPUT+"users_huc8_4",
                          out_feature_class=OUTPUT+"users_huc8_5",
                          dissolve_field=["user_id", "HUC_8"], statistics_fields="pwsid FIRST; username FIRST; source FIRST; AGENCYUNIQ FIRST; user_x FIRST; user_y FIRST; totarea FIRST; huc8_area FIRST; huc8_pctcrop FIRST",
                          multi_part="MULTI_PART", unsplit_lines="DISSOLVE_LINES")
print "Dissolving (2/2)..."
arcpy.Dissolve_management(in_features=OUTPUT+"users_dauco_4",
                          out_feature_class=OUTPUT+"users_dauco_5",
                          dissolve_field=["user_id", "dauco_id"], statistics_fields="pwsid FIRST; username FIRST; source FIRST; AGENCYUNIQ FIRST; user_x FIRST; user_y FIRST; totarea FIRST; DAU_CODE FIRST; DAU_NAME FIRST; PSA_CODE FIRST; PSA_NAME FIRST; HR_CODE FIRST; HR_NAME FIRST; PA_NO FIRST; COUNTY_NAME FIRST; COUNTY_CODE FIRST; COUNTY_ANSI FIRST; dauco_area FIRST; dauco_pctcrop FIRST",
                          multi_part="MULTI_PART", unsplit_lines="DISSOLVE_LINES")
print "Renaming variables..."
huc8_vars = ["pwsid", "username", "source", "AGENCYUNIQ", "user_x", "user_y", "totarea", "huc8_area", "huc8_pctcrop"]
dauco_vars  = ["pwsid", "username", "source", "AGENCYUNIQ", "user_x", "user_y", "totarea", "DAU_CODE", "DAU_NAME", "PSA_CODE", "PSA_NAME", "HR_CODE", "HR_NAME", "PA_NO", "COUNTY_NAME", "COUNTY_CODE", "COUNTY_ANSI", "dauco_area", "dauco_pctcrop"]
for col in huc8_vars:
    arcpy.AlterField_management(in_table=OUTPUT+"users_huc8_5", field="FIRST_"+col, new_field_name=col)
for col in dauco_vars:
    arcpy.AlterField_management(in_table=OUTPUT+"users_dauco_5", field="FIRST_"+col, new_field_name=col)

# Add unique id
arcpy.AddField_management(in_table=OUTPUT+"users_huc8_5", field_name="iuser_id", field_type="LONG")
arcpy.CalculateField_management(in_table=OUTPUT+"users_huc8_5", field="iuser_id", expression="!OBJECTID!", expression_type="PYTHON_9.3")
arcpy.AddField_management(in_table=OUTPUT+"users_dauco_5", field_name="iuser_id", field_type="LONG")
arcpy.CalculateField_management(in_table=OUTPUT+"users_dauco_5", field="iuser_id", expression="!OBJECTID!", expression_type="PYTHON_9.3")

# Add centroids within intersected users (restricting to within shape)
print "Calculating centroids..."
arcpy.AddGeometryAttributes_management(OUTPUT+"users_huc8_5", "CENTROID_INSIDE", "", "", coordsys_geog)
arcpy.AddGeometryAttributes_management(OUTPUT+"users_dauco_5", "CENTROID_INSIDE", "", "", coordsys_geog)
arcpy.AlterField_management(in_table=OUTPUT+"users_huc8_5", field="INSIDE_X", new_field_name="iuser_x")
arcpy.AlterField_management(in_table=OUTPUT+"users_huc8_5", field="INSIDE_Y", new_field_name="iuser_y")
arcpy.AlterField_management(in_table=OUTPUT+"users_dauco_5", field="INSIDE_X", new_field_name="iuser_x")
arcpy.AlterField_management(in_table=OUTPUT+"users_dauco_5", field="INSIDE_Y", new_field_name="iuser_y")
arcpy.DeleteField_management(OUTPUT+"users_huc8_5", "INSIDE_Z")
arcpy.DeleteField_management(OUTPUT+"users_huc8_5", "INSIDE_M")
arcpy.DeleteField_management(OUTPUT+"users_dauco_5", "INSIDE_Z")
arcpy.DeleteField_management(OUTPUT+"users_dauco_5", "INSIDE_M")

# Add shape areas for intersected users
print "Calculating areas..."
arcpy.AddGeometryAttributes_management(OUTPUT+"users_huc8_5", "AREA", "", "SQUARE_KILOMETERS")
arcpy.AlterField_management(in_table=OUTPUT+"users_huc8_5", field="POLY_AREA", new_field_name="iuser_area")
arcpy.AddGeometryAttributes_management(OUTPUT+"users_dauco_5", "AREA", "", "SQUARE_KILOMETERS")
arcpy.AlterField_management(in_table=OUTPUT+"users_dauco_5", field="POLY_AREA", new_field_name="iuser_area")

# Calculate zonal statistics on cropmask and join back
print "Calculating cropmask (1/2 - may take up to an hour)..."
arcpy.ZonesWOverlap(Zonal_Feature_Class="users_huc8_5",
                    Zone_Field="iuser_id",
                    Value_Raster=RAW_CROPMASK,
                    Workspace="H:/analysis/ca/gis/scratch",
                    Zonal_Statistics_Table=OUTPUT+"users_huc8_cropmask",
                    Feature_Class_Divides="10")
print "Calculating cropmask (2/2 - may take up to an hour)..."
arcpy.ZonesWOverlap(Zonal_Feature_Class="users_dauco_5",
                    Zone_Field="iuser_id",
                    Value_Raster=RAW_CROPMASK,
                    Workspace="H:/analysis/ca/gis/scratch",
                    Zonal_Statistics_Table=OUTPUT+"users_dauco_cropmask",
                    Feature_Class_Divides="10")
arcpy.JoinField_management(in_data=OUTPUT+"users_huc8_5", in_field="iuser_id",
                           join_table=OUTPUT+"users_huc8_cropmask", join_field="IUSER_ID", fields="MEAN")
arcpy.JoinField_management(in_data=OUTPUT+"users_dauco_5", in_field="iuser_id",
                           join_table=OUTPUT+"users_dauco_cropmask", join_field="IUSER_ID", fields="MEAN")
arcpy.AlterField_management(in_table=OUTPUT+"users_huc8_5", field="MEAN", new_field_name="iuser_pctcrop")
arcpy.AlterField_management(in_table=OUTPUT+"users_dauco_5", field="MEAN", new_field_name="iuser_pctcrop")

# Export tables
arcpy.TableToExcel_conversion(Input_Table=OUTPUT+"users_huc8_5", Output_Excel_File=FINAL_TABLES+"users_huc8.xls")
arcpy.TableToExcel_conversion(Input_Table=OUTPUT+"users_dauco_5", Output_Excel_File=FINAL_TABLES+"users_dauco.xls")
arcpy.TableToExcel_conversion(Input_Table=RAW_ATLAS_FEDERAL_TABLE, Output_Excel_File=FINAL_TABLES+"federal_table.xls")


except:
    print arcpy.GetMessages()
    raise    

File: R/Codes/GIS/make_users.R
library(readr)
library(readxl)
library(foreign)
library(sf)
library(stringi)
library(ggplot2)
library(Hmisc)
library(raster)
library(rgdal)
library(tidyverse)
library(dplyr)
library(units)
library(exactextractr)

rm(list=ls())

##Setting_directories (Change root directory while running codes)

Root<- "/Users/shubhalakshminag/Box/CA_Surface_water/"

##GIS_DATA

DATA_GIS<-paste0(Root, "data/gis/")

##FINAL_OUTPUT

OUTPUT_GIS<-paste0( Root, "R/GIS/")

#IMPORTING SHAPE FILES

RAW_COUNTY<-st_read(paste0(DATA_GIS, "ca_atlas/counties/cnty24k09_1_multipart.shp"))
RAW_DAU<-st_read(paste0(DATA_GIS, "dwr/dau/dau_v2_105.shp"))
RAW_DWR_AGENCIES<-st_read(paste0(DATA_GIS, "dwr/districts/combined/water_agencies.shp"))
RAW_ATLAS_FEDERAL<-st_read(paste0(DATA_GIS, "ca_atlas/districts/federal/FederalWaterUsers.shp"))
RAW_ATLAS_STATE<-st_read(paste0(DATA_GIS, "ca_atlas/districts/state/wdst24.shp"))
RAW_ATLAS_PRIVATE<-st_read(paste0(DATA_GIS, "ca_atlas/districts/private/wdpr24.shp"))
RAW_MOJAVE<-st_read(paste0(DATA_GIS, 
                           "ca_atlas/mojave/Mojave_Water_Agency_Service_Area_Water_Companies_2012.shp"))
RAW_CEHTP<-st_read(paste0(DATA_GIS, "cehtp/service_areas.shp"))
RAW_CROPMASK<-raster(paste0(DATA_GIS, "nass_cdl/CMASK_2015_06.tif"))
RAW_ATLAS_FEDERAL_TABLE<-mdb.get("WD-WaterUsers.mdb", "Master")
huc8_1<-st_read(paste0(DATA_GIS, "dwr/watersheds/huc8_1.shp"))

#Dissolve within layer, into multipart features

agencies_1<-RAW_DWR_AGENCIES%>%
  st_buffer(0)%>%
  group_by(AGENCYUNIQ)%>%
  summarise(AGENCYNAME=first(AGENCYNAME))%>%
  ungroup()

federal_1<-RAW_ATLAS_FEDERAL%>%
  st_buffer(0)%>%
  group_by(WDNAME)%>%
  summarise()%>%
  ungroup()%>%
  st_cast(., "MULTIPOLYGON")

swp_1<-RAW_ATLAS_STATE%>%
  group_by(WDNAME)%>%
  summarise()%>%
  ungroup()

private_1<-RAW_ATLAS_PRIVATE%>%
  group_by(WDNAME)%>%
  summarise()%>%
  ungroup()

mojave_1<-RAW_MOJAVE%>%
  group_by(Name)%>%
  summarise()%>%
  ungroup()

cehtp_1<-RAW_CEHTP%>%
  group_by(pwsid)%>%
  summarise(pwsname=first(pwsname))%>%
  ungroup()

RAW_DAU<-st_transform(RAW_DAU, 3310)

dau_1<-RAW_DAU%>%
  group_by(DAU_CODE)%>%
  summarise(DAU_NAME=first(DAU_NAME),
            PSA_CODE=first(PSA_CODE),
            PSA_NAME=first(PSA_NAME),
            HR_CODE=first(HR_CODE),
            HR_NAME=first(HR_NAME),
            PA_NO=first(PA_NO))%>%
  ungroup()

#Intersect DAU and county

dauco_1<-dau_1%>%
  st_buffer(-215)%>%
  st_intersection(RAW_COUNTY)

#Deleting unnecessary features

dauco_1<-dauco_1%>%
  dplyr::select(-c(NAME_UCASE, FMNAME_PC, FMNAME_UC, ABBREV, ABCODE))%>%
  arrange(DAU_CODE)

#Renaming features

agencies_1<-agencies_1%>%
  rename(username=AGENCYNAME)

federal_1<-federal_1%>%
  rename(username=WDNAME)

private_1<-private_1%>%
  rename(username=WDNAME)

mojave_1<-mojave_1%>%
  rename(username=Name)

cehtp_1<-cehtp_1%>%
  rename(username=pwsname)

swp_1<-swp_1%>%
  rename(username=WDNAME)

dauco_1<-dauco_1%>%
  rename(COUNTY_NAME=NAME_PCASE,
         COUNTY_CODE=NUM,
         COUNTY_ANSI=ANSI)

#Adding a source variable 

agencies_1<-agencies_1%>%
  mutate(source="agencies")

federal_1<-federal_1%>%
  mutate(source="federal")

swp_1<-swp_1%>%
  mutate(source="swp")

private_1<-private_1%>%
  mutate(source="private")

mojave_1<-mojave_1%>%
  mutate(source="mojave")

cehtp_1<-cehtp_1%>%
  mutate(source="cehtp")

#Add a field identifying DAUCOid

dauco_1<-dauco_1%>%
  mutate(dauco_id=sub("^0+", "",DAU_CODE ),
         dauco_id=as.numeric(dauco_id)*100 + COUNTY_CODE)

#Merge users together (projecting into WGS84, the first input's coordinate system)  
#Transform all other datasets to crs=4326 (projecting into WGS84)

cehtp_1<-st_transform(cehtp_1, 4326)
agencies_1<-st_transform(agencies_1, 4326)
st_crs(federal_1)<-26911
federal_1<-st_transform(federal_1, 4326)
swp_1<-st_transform(swp_1, 4326)
private_1<-st_transform(private_1, 4326)
mojave_1<-st_transform(mojave_1, 4326)

## Add centroid for mojave

mojave_1<-mojave_1%>%
  st_buffer(0)%>%
  st_zm()

mojave_1$centroid<-mojave_1%>%
  st_point_on_surface()%>%
  st_geometry()

mojave_1<-mojave_1%>%
  mutate(user_x=sf::st_coordinates(centroid)[,1],
         user_y=sf::st_coordinates(centroid)[,2])

# Merging users

users_2<-bind_rows(cehtp_1, agencies_1, federal_1, swp_1, private_1)

#Adding centroids (restricting to points within the polygons)

users_2$centroid<-users_2%>%
  st_point_on_surface()%>%
  st_geometry()

users_3<-users_2%>%
  mutate(user_x=sf::st_coordinates(centroid)[,1],
         user_y=sf::st_coordinates(centroid)[,2])

# Add user unique id

users_3<-users_3%>%
  bind_rows(mojave_1)%>%
  mutate(user_id=row_number())%>%
  filter(!st_is_empty(.))

# Project to meters and calculate area-change crs to 26919 ('NAD 1983 UTM Zone 10N')

users_3<-st_transform(users_3, crs=26911 )
huc8_2<-st_transform(huc8_1, crs=26911 )
dauco_2<-st_transform(dauco_1, crs=26911 )

# Adding Shape Areas

users_3<-users_3%>%
  mutate(totarea=sf::st_area(.),
         totarea=set_units(totarea, km^2))

huc8_2<-huc8_2%>%
  mutate(huc8area=sf::st_area(.),
         huc8area=set_units(huc8area, km^2))

dauco_2<-dauco_2%>%
  mutate(daucoarea=sf::st_area(.),
         daucoarea=set_units(daucoarea, km^2))

# Calculate zonal statistics on cropmask and join back

huc8_pctcrop<-as.data.frame(exact_extract(RAW_CROPMASK, huc8_2, 'mean'))
huc8_pctcrop$HUC_8=huc8_2$HUC_8

huc8_2<-huc8_2%>%
  inner_join(huc8_pctcrop)%>%
  rename("huc8_pctc"="exact_extract(RAW_CROPMASK, huc8_2, \"mean\")")

dauco_pctcrop<-as.data.frame(exact_extract(RAW_CROPMASK, dauco_2, 'mean'))
dauco_pctcrop$dauco_id=dauco_2$dauco_id

dauco_2<-dauco_2%>%
  inner_join(dauco_pctcrop)%>%
  rename("dauco_pctc"="exact_extract(RAW_CROPMASK, dauco_2, \"mean\")")

users_final<-st_transform(users_3, 4326)%>%
  dplyr::select(-centroid)%>%
  mutate(AGENCYUNIQ=ifelse(is.na(AGENCYUNIQ), 0, AGENCYUNIQ))
huc8_final<-st_transform(huc8_2, 4326)
dauco_final<-st_transform(dauco_2, 4326)

save(huc8_final, file = paste0(OUTPUT_GIS, "huc8_final.RData"))
save(dauco_final, file = paste0(OUTPUT_GIS, "dauco_final.RData"))
save(users_final, file = paste0(OUTPUT_GIS, "users_final.RData"))

# Intersect users with DAUCo and HUC8

users_huc8_4<-users_3%>%
  st_buffer(440)%>%
  st_intersection(huc8_2)

users_dauco_4<-users_3%>%
  st_buffer(393)%>%
  st_intersection(dauco_2)

#Dissolving polygons

users_huc8_5<-users_huc8_4%>%
  group_by(user_id, HUC_8)%>%
  summarise(pwsid=first(pwsid),
            username=first(username),
            source=first(source),
            AGENCYUNIQ=first(AGENCYUNIQ),
            user_x=first(user_x),
            user_y=first(user_y),
            totarea=first(totarea),
            huc8_area=first(huc8area),
            huc8_pctc=first(huc8_pctc))%>%
  ungroup()

users_dauco_5<-users_dauco_4%>%
  group_by(user_id, dauco_id)%>%
  summarise(pwsid=first(pwsid),
            username=first(username),
            source=first(source),
            AGENCYUNIQ=first(AGENCYUNIQ),
            user_x=first(user_x),
            user_y=first(user_y),
            totarea=first(totarea),
            daucoarea=first(daucoarea),
            dauco_pctc=first(dauco_pctc),
            DAU_CODE=first(DAU_CODE),
            DAU_NAME=first(DAU_NAME),
            PSA_CODE=first(PSA_CODE),
            PSA_NAME=first(PSA_NAME),
            HR_CODE=first(HR_CODE),
            HR_NAME=first(HR_NAME),
            PA_NO=first(PA_NO),
            COUNTY_NAME=first(COUNTY_NAME),
            COUNTY_CODE=first(COUNTY_CODE),
            COUNTY_ANSI=first(COUNTY_ANSI))%>%
  ungroup()

#Adding unique ID

users_huc8_5<-users_huc8_5%>%
  mutate(iuser_id=row_number())

users_dauco_5<-users_dauco_5%>%
  mutate(iuser_id=row_number())

# Add centroids within intersected users (restricting to within shape)

users_huc8_5<-st_transform(users_huc8_5,4326)
users_huc8_5$centroid<-users_huc8_5%>%
  st_buffer(0)%>%
  st_point_on_surface()%>%
  st_geometry()
users_huc8_5<-users_huc8_5%>%
  mutate(iuser_x=sf::st_coordinates(centroid)[,1],
         iuser_y=sf::st_coordinates(centroid)[,2])

users_dauco_5<-st_transform(users_dauco_5,4326)
users_dauco_5$centroid<-users_dauco_5%>%
  st_buffer(0)%>%
  st_point_on_surface()%>%
  st_geometry()
users_dauco_5<-users_dauco_5%>%
  mutate(iuser_x=sf::st_coordinates(centroid)[,1],
         iuser_y=sf::st_coordinates(centroid)[,2])

# Add shape areas for intersected users

users_dauco_5<-st_transform(users_dauco_5,26911)
users_dauco_5<-users_dauco_5%>%
  mutate(iuser_area=sf::st_area(.),
         iuser_area=set_units(iuser_area, km^2))

users_huc8_5<-st_transform(users_huc8_5,26911)
users_huc8_5<-users_huc8_5%>%
  mutate(iuser_area=sf::st_area(.),
         iuser_area=set_units(totarea, km^2))

# Calculate zonal statistics on cropmask and join back

user_huc8_pctcrop<-as.data.frame(exact_extract(RAW_CROPMASK, users_huc8_5, 'mean'))
user_huc8_pctcrop$iuser_id=users_huc8_5$iuser_id

users_huc8_5<-users_huc8_5%>%
  inner_join(user_huc8_pctcrop)%>%
  rename("iuser_pctc"="exact_extract(RAW_CROPMASK, users_huc8_5, \"mean\")")

user_dauco_pctcrop<-as.data.frame(exact_extract(RAW_CROPMASK, users_dauco_5, 'mean'))
user_dauco_pctcrop$iuser_id=users_dauco_5$iuser_id

users_dauco_5<-users_dauco_5%>%
  inner_join(user_dauco_pctcrop)%>%
  rename("iuser_pctc"="exact_extract(RAW_CROPMASK, users_dauco_5, \"mean\")")

# Saving final data

save(users_dauco_5, file = paste0(OUTPUT_GIS, "users_dauco.RData"))
save(users_huc8_5, file = paste0(OUTPUT_GIS, "users_huc8.RData"))
save(RAW_ATLAS_FEDERAL_TABLE, file = paste0(OUTPUT_GIS, "federal_table.xlsx"))






